flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/co/CoBroadcastWithKeyedOperatorTest.java

```diff
@@ -0,0 +1,734 @@
+
+		private static final long serialVersionUID = 7496674620398203933L;
+
+		private final Set<String> keysToRegister;
+
+		TestFunctionWithOutput(Set<String> keysToRegister) {
+			this.keysToRegister = Preconditions.checkNotNull(keysToRegister);
+		}
+
+		@Override
+		public void processElementOnBroadcastSide(Integer value, KeyedReadWriteContext ctx, Collector<String> out) throws Exception {
+			// put an element in the broadcast state
+			for (String k : keysToRegister) {
+				ctx.putToBroadcast(k, value);
+			}
+		}
+
+		@Override
+		public void processElement(String value, KeyedReadOnlyContext ctx, Collector<String> out) throws Exception {
+			for (Map.Entry<String, Integer> entry : ctx.readOnlyBroadcastIterable()) {
+				out.collect(entry.toString());
+			}
+		}
+	}
+
+	@Test
+	public void testNoKeyedStateOnBroadcastSide() throws Exception {
+
+		boolean exceptionThrown = false;
+
+		try (
+				AutoClosableTestHarness<String, Integer, String, String, Integer, String> autoTestHarness = new AutoClosableTestHarness<>(
+						BasicTypeInfo.STRING_TYPE_INFO,
+						new IdentityKeySelector<>(),
+						new KeyedBroadcastProcessFunction<String, Integer, String, String, Integer, String>() {
+
+							private static final long serialVersionUID = -1725365436500098384L;
+
+							private final ValueStateDescriptor<String> valueState = new ValueStateDescriptor<>("any", BasicTypeInfo.STRING_TYPE_INFO);
+
+							@Override
+							public void processElementOnBroadcastSide(Integer value, KeyedReadWriteContext ctx, Collector<String> out) throws Exception {
+								getRuntimeContext().getState(valueState).value(); // this should fail
+							}
+
+							@Override
+							public void processElement(String value, KeyedReadOnlyContext ctx, Collector<String> out) throws Exception {
+								// do nothing
+							}
+						},
+						BasicTypeInfo.STRING_TYPE_INFO,
+						BasicTypeInfo.INT_TYPE_INFO)
+		) {
+			TwoInputStreamOperatorTestHarness<Integer, String, String> testHarness = autoTestHarness.getTestHarness();
+
+			testHarness.processWatermark1(new Watermark(10L));
+			testHarness.processWatermark2(new Watermark(10L));
+			testHarness.processElement1(new StreamRecord<>(5, 12L));
+		} catch (NullPointerException e) {
+			Assert.assertEquals("No key set. This method should not be called outside of a keyed context.", e.getMessage());
+			exceptionThrown = true;
+		}
+
+		if (!exceptionThrown) {
+			Assert.fail("No exception thrown");
+		}
+	}
+
+	private static class IdentityKeySelector<T> implements KeySelector<T, T> {
+		private static final long serialVersionUID = 1L;
+
+		@Override
+		public T getKey(T value) throws Exception {
+			return value;
+		}
+	}
+
+	/**
+	 * A wrapper of the test harness that makes sure to close it after the test finishes.
+	 */
+	private static class AutoClosableTestHarness<KEY, IN1, IN2, K, V, OUT> implements AutoCloseable {
```

Comment:
The harnesses themselves are already `AutoCloseable`. (It's a somewhat newer addition, though.)